Introduce placeholder express methods

maxparkergene opened the pull request:
Previously there was a large CSL score because all of the Adyen Express methods were being lazy loaded and only showing after the network request had finished.

The BlueFinch Checkout has had an update allowing for placeholder express methods to be rendered and remain visible until the lazy loading of the full Express methods.

This update introduces these placeholders and then switches them out after we have fully rendered the express methods.

<!-- 

Do not commit any changes to the `view/frontend/web/js/checkout/dist` directory 

See .github/CONTRIBUTING.md for local workflow recommendations

-->

maxparkergene committed:
Introduce placeholder express methods
Previously there was a large CSL score because all of the Adyen Express methods were being lazy loaded and only showing after the network request had finished.

The BlueFinch Checkout has had an update allowing for placeholder express methods to be rendered and remain visible until the lazy loading of the full Express methods.

This update introduces these placeholders and then switches them out after we have fully rendered the express methods.

maxparkergene committed:
Added Adyen requirement

convenient commented on Plugin/CustomerData/Cart.php:
```diff
@@ -0,0 +1,58 @@
+                $this->adyenPaymentMethodManagement->getPaymentMethods($quoteId),
+                true
+            ) : [];
+        $result[self::PAYMENT_METHODS_KEY] = $paymentMethods;
```
@maxparkergene This will currently fire after all section data loads, but if adyen express payments is enabled we're already doing this request once. This will help reduce load on adyen and keep this lean

![Screenshot 2025-05-23 at 14 27 14](https://github.com/user-attachments/assets/7baeb86b-f04b-4824-8d42-2e07175a3ff6)

We need to handle two cases
- Adyen express enabled
   - we need to check storeConfigPaths to see if express is active and enabled
   - if it is, then we terminate this plugin because we dont want to duplicate api requests
   - we also need a new plugin or extension, after the adyen express functionality to grab that payload and use it for our purposes. so that we dont run the same request twice
- Adyen express not enabled
   - use this implementation, as its very similar to the adyen express method.

maxparkergene replied:
Hi @convenient,

Thanks for the feedback and I have added the checks as you've suggested to only run the Adyen API request if the express module is not enabled. If it is there is a separate plugin that will take the data and use it as we are expecting in BlueFinch checkout.

maxparkergene committed:
Creation of checks to prevent calling Adyen API multiple times
If the Adyen Express module is enabled then they will have already gathered the payment methods to be added into the customer data.

Added a check to see if any of the configuration for Adyen Express is enabled and if so we will not run the lookup ourselves but instead use a plugin to take the data that already exists and use it for our purposes.

maxparkergene committed:
Fix issue with failing phpstan tests

maxparkergene committed:
Fix issue with version of Magento Coding Standards

tr33m4n commented on Plugin/CustomerData/Cart.php:
```diff
@@ -0,0 +1,29 @@
+        Subject $subject,
+        array $result
+    ): array {
+        if (isset($result[Subject::PAYMENT_METHODS_KEY])) {
```
Hey @maxparkergene, I might have missed something here, but looking at the Adyen plugin https://github.com/Adyen/adyen-magento2-express-checkout/blob/main/Plugin/CustomerData/Cart.php#L81 it looks like payment methods are loaded no matter what. Does that make the check here https://github.com/BlueFinchCommerce/module-checkout-adyen/pull/20/changes#diff-8953722ac5f9475cb6efa8979d7f7d23737440d0b5b35ad663f785727e208f52R64 redundant?

maxparkergene replied:
Hi @tr33m4n, the reason we added this in was for the case where a merchant has the Adyen module installed but not the Adyen express module. If they are missing the express module then this would come in and add in the required section data that would have been missing otherwise.

tr33m4n commented on Plugin/CustomerData/AddAdyenExpressMethods.php:
```diff
@@ -0,0 +1,76 @@
+        $payPalExpressEnabled = $this->scopeConfig->getValue('payment/adyen_paypal_express/express_show_on', ScopeInterface::SCOPE_STORE);
+
+        // Only run if no Adyen Express methods are enabled otherwise we'd be duplicating the Adyen API request.
+        if (!$googlePayExpressEnabled && !$applePayExpressEnabled && !$payPalExpressEnabled) {
```
@maxparkergene As the package can be installed and potentially not configured, we could see duplicate requests just with the express module sitting in the background. The better check here might be something using the `\Magento\Framework\Module\Manager` like:
```php
if (!$this->moduleManager->isEnabled('Adyen_ExpressCheckout')) {
    //
}

tr33m4n reviewed (changes requested):
@maxparkergene Looking good, added a comment about idle packages

maxparkergene committed:
Update to use Module\Manager to determine Adyen_ExpressCheckout enabled state
